### PR TITLE
Add mutual friendship rows

### PR DESCRIPTION
**Pull request by arikarim**
Added associations to facilitate friendships between users.
users are able to send a request and also accept or deny requests.
mutual friends are displayed next to user names.

### CODE REVIEW

**Comment by arikarim** [comment]
Hello guys.
you don't see a lot of commits because we did most of the things in the previous milestone which just been approved.

please have a review and thanks for your hard work:)).

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @arikarim and @apricot12, both of you have done an awesome job so far 👌. 

* Excellent use of capybara for your integration tests 👏
* Great use of helper methods to strip logic from your views 💪 

There are a couple of changes that are necessary for this project to be approved.

I see you have implemented most, if not all of the required feature, however for our application to scale properly, kindly consider the following and update yours accordingly.

Considering the “Fat Model, Skinny Controller” approach, this refers to how the M and C parts of MVC ideally work together. Namely, any non-response-related logic should go in the model, ideally in a nice, testable method. Meanwhile, the “skinny” controller is simply a nice interface between the view and model.

- [x] Your `#confirm` method inside the Users controller does the job, however it can be improved, since it will not be involving any need to use any Ruby method which adds to the complexity of the method.
```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] Thanks to the above method, we can find user friends only by user_id in friendships table
```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```
With this method you will no longer need the `#friends` Ruby method

- [x] User `#pending_friends` can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```
With this method you will no longer need the `#pending_requests` Ruby method

- [x] User `#friend_requests` can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

With this method you will no longer need the `#friend_requests` Ruby method

- [x] In order to get all posts written by user on their friends. I know this part works well in your app, but following MVC pattern we try as much as possible to keep all business logic in models and let controller just do the work of connecting the models and views. 
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
With this method you will simply have
```ruby
  def timeline_posts
    @timeline_posts ||= current_user.friends_and_own_posts.ordered_by_most_recent.includes(:user)
  end
```

<hr />

- [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
See affected file below:
> `layouts/application.html.erb`
> `posts/_posts.html.erb`
> `users/_users.html.erb`

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

<hr />

- [x] Your ReadMe file still carries the app template description. "This repo includes initial code for social media app with basic styling. Its purpose is to be a starting point for Microverse students." I am sure you will agree with me, this repo no longer carries initial code, you have done a lot. Do well to update it to reflect what the app is all about. Possibly a screenshot  and author names too, to give it that professional look.

### Tests
- [x] You did a great work with the feature  specs, but the we need to have unit tests to cover every model we have in you app. Currently you only have unit tests for friendship model, add for that other ones.


### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Comment by arikarim** [comment]
Thanks for the review it was really helpful.

kindly have a look.

File: app/models/user.rb [hunk 1]
```diff
@@ -9,27 +7,18 @@ class User < ApplicationRecord
...
-    [friends_array1, friends_array2].flatten
-  end
+  has_many :confirmed_friendships, -> { where status: true }, class_name: 'Friendship'
+  has_many :friends, through: :confirmed_friendships
 
-  def pending_requests
-    friendships.map { |friendship| friendship.friend unless friendship.status }.compact
-  end
+  has_many :pending_friendships, -> { where status: false }, class_name: 'Friendship', foreign_key: 'user_id'
+  has_many :pending_requests, through: :pending_friendships, source: :friend
```
Comment (enaburekhan) [inline app/models/user.rb]: Replace `pending_requests` with `pending_friends`

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
Hi team!
Great job so far:clap::clap:. Thank you for implementing most of the changes requested by the last reviewer. The readme is now very descriptive:100:.
However, we still need to make a few changes to get the milestone approved:muscle:.

**To improve**
As highlighted by the last reviewer:
```ruby
# Friendship model
def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end

- [ ] thanks to this, we can find user friends only by user_id in friendships table

```

- [x] This is still missing in the friendship model. Kindly implement it.

- [x] As highlighted in the PR below. Correct `pending_requests` to `pending_friends` in user model.
- [x] Move the logic in `application.html.erb`, `_post.html.erb`, and` _user.html.erb` to a helper.  There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. The last reviewer provided a resource on how to use a helper in rails. Kindly use it if you need help.

Make the required changes and when you are ready, request another review.

Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan

**Comment by arikarim** [comment]
Thanks for the review 
please have another look

**Review by voscarmv — APPROVED** [review]
# Project status: Approved 💯 ✨ 😎 👍 

- Great job applying the changes requested by the previous Code Reviewers.

### You may merge this branch and move forward ⛙🛣️ 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻
# Happy Coding! 🧠 ⌨️ 🖱️ 🖥️

**Review by MahmoudBakr23 — COMMENTED** [review]
# **Changes Required 🛑♻**

<hr>

Hello there :)
Congratulations on getting to this final step of the project. Valid live demo link and descriptive README file.
Keep up the great work :)

- [x] Please consider create a PR that asks for merging from the development branch to the master branch.

<hr>

Good luck with the next and happy coding 💻👨‍💻
Please feel free to contact me on Slack with the name "Mahmoud Bakr" if you had any questions.

**Pull request merged**